notifications_utils/postal_address.py

```diff
@@ -80,6 +81,12 @@ def has_too_many_lines(self):
     def has_valid_postcode(self):
         return self.postcode is not None
 
+    @property
+    def has_valid_last_line(self):
+        return self.has_valid_postcode or (
```

Comment:
i think the checks should be flipped round for marginal performance gainz: `return (self.internationl_letters and self.international) or self.has_valid_postcode`

`has_valid_postcode` performs regex checks and string formatting, while `self.international_letters` is a simple boolean while `self.international` creates a new Country object each time, but at least the country mapping underneath is lru_cache'd.

 (nb shoult `has_valid_postcode` be cached? an lru_cache probs won't do too much since most postcodes will be different, but should this be stored in a private var on self (like `self._has_valid_postcode`?)

Reply:
Flipped the order is cc8bfcd – makes sense.

I don’t think it’s worth per-instance caching `has_valid_postcode`. For CSVs we’ll call it at most once per row and then once for the first 50 rows to work out which error message to show.

And I’m kind of wary about caching recipient validation generally because it could hide a real life performance problem which the tests don’t catch because they test against `"a@b.com\n" * RecipientCSV.max_rows`